oeps/oep-0006.rst

```diff
@@ -0,0 +1,224 @@
+
+The existing named scopes, and an example use case for each, are as follows:
+
+* ``Scope.content`` (UserScope.NONE, BlockScope.DEFINITION):
+
+  * Example: The text of a multiple choice question
+* ``Scope.settings`` (UserScope.NONE, BlockScope.USAGE)
+
+  * Example: The number of attempts allowed for a problem
+* ``Scope.user_state`` (UserScope.ONE, BlockScope.USAGE)
+
+  * Example: The student's answer to a multiple choice question
+* ``Scope.user_state_summary`` (UserScope.ALL, BlockScope.USAGE)
+
+  * Example: The number of students who have answered the question (side note: there is no locking or atomicity for XBlock field writes, so in practice this field is currently of little use in a MOOC)
+* ``Scope.user_info`` (UserScope.ONE, BlockScope.ALL)
+
+  * Example: user age (note: this scope seems little-used and overlaps with the user service)
+* ``Scope.preferences`` (UserScope.ONE, BlockScope.TYPE)
+
+  * Example: (For a video XBlock) Boolean setting to prefer YouTube or non-YouTube video playback.
+
+
+Motivation
+==========
+
+As one would expect, ``Scope.user_info`` and ``Scope.preferences`` fields in the Open edX platform are implemented globally, and not scoped to the course that uses them. There is currently no standard, extensible way to use XBlock fields to store data that is scoped to a particular "context," such as a course, although this is a common need.
+
+
+Specification
+=============
+
+New BlockScope values
+---------------------
+
+To support context-scoped fields, ``BlockScope`` will allow these new additional values:
+
+* ``CONTEXT_TYPE``: The data is related to all instances of this type of XBlock within a particular context.
+* ``CONTEXT_ALL``: The data is accessible to XBlocks of any type, within a particular context. (As with ``BlockScope.ALL``, XBlock authors who use such fields should be careful to use a namespace that will not conflict with ``BlockScope.CONTEXT_ALL`` fields defined by other XBlock authors.)
+
+A "context" is defined by the runtime but is usually a "course" or the nearest equivalent. In the Open edX platform, the context may be a course or a content library. Different runs of the same course would be considered different contexts and would not share data, though UserScope.NONE fields would be copied when creating a new course run. On other platforms, a "context" could be defined as a blog post or a textbook.
+
+New named scopes
+----------------
+
+As is the situtation today, XBlock runtimes would generally not be expected to implement every possible combination of ``BlockScope`` and ``UserScope``. Instead, XBlock runtimes are encouraged to implement these new named scopes, which would be defined in the XBlock python package:
+
+* ``Scope.context_settings`` (UserScope.NONE, BlockScope.CONTEXT_TYPE)
+
+  * Example: For an LTI XBlock, a list of the LTI provider URLs and secrets used in the course
+  * Example: For a discussions XBlock: Enable or disable discussions in this course
```

Comment:
With the (now deprecated) mentoring XBlock, we [used](https://github.com/edx-solutions/xblock-mentoring/blob/f461107/mentoring/models.py) a custom django model to store answers so that authors could use that XBlock to prompt students for a free text response answer early in the course, then use a separate instance of that XBlock later in the course to show the student's earlier answer and give them the opportunity to revise it. This approach required that `'mentoring'` be added to the platform's `OPTIONAL_APPS` setting, which is not a pluggable/scalable solution.

For its successor, the Problem Builder suite of XBlocks, we use `edx-submissions` to store answers, and that approach also allows this sort of data sharing among XBlocks (and is a django model). The advantage of that is that it doesn't require any change to `OPTIONAL_APPS`.